components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesBookmarksStorage.kt

```diff
@@ -0,0 +1,177 @@
+import mozilla.appservices.places.BookmarkFolder
+import mozilla.appservices.places.BookmarkItem
+import mozilla.appservices.places.BookmarkRoot
+import mozilla.appservices.places.BookmarkSeparator
+import mozilla.appservices.places.BookmarkTreeNode
+import mozilla.appservices.places.BookmarkUpdateInfo
+import mozilla.appservices.places.PlacesException
+import mozilla.appservices.places.PlacesReaderConnection
+import mozilla.appservices.places.PlacesWriterConnection
+import mozilla.components.concept.storage.BookmarksStorage
+import mozilla.components.concept.sync.AuthInfo
+import mozilla.components.concept.sync.SyncStatus
+import mozilla.components.concept.sync.SyncableStore
+
+/**
+ * Implementation of the [BookmarksStorage] which is backed by a Rust Places lib via [PlacesConnection].
+ */
+open class PlacesBookmarksStorage(context: Context) : BookmarksStorage, SyncableStore {
+
+    private val scope by lazy { CoroutineScope(IO) }
+    private val storageDir by lazy { context.filesDir }
+    @VisibleForTesting
+    internal open val places: Connection by lazy {
+        RustPlacesConnection.init(storageDir)
+        RustPlacesConnection
+    }
+
+    private val writer: PlacesWriterConnection by lazy { places.writer() }
+    private val reader: PlacesReaderConnection by lazy { places.reader() }
+
+    /**
+     * Produces a populated bookmarks tree for the given root.
+     *
+     * @param root The bookmark root to obtain.
+     * @param recursive Whether to recurse and obtain all levels of children.
+     * @return The populated root.
+     */
+    override suspend fun getTree(root: BookmarkRoot, recursive: Boolean): BookmarkTreeNode? {
+        return withContext(scope.coroutineContext) {
+            reader.getBookmarksTree(root.id, recursive)
+        }
+    }
+
+    /**
+     * Produces a populated bookmarks tree for the given guid string.
+     *
+     * @param guid The bookmark guid to obtain.
+     * @param recursive Whether to recurse and obtain all levels of children.
+     * @return The populated root starting from the guid.
+     */
+    override suspend fun getTree(guid: String, recursive: Boolean): BookmarkTreeNode? {
+        return withContext(scope.coroutineContext) {
+            reader.getBookmarksTree(guid, recursive)
+        }
+    }
+
+    /**
+     * Searches bookmarks with a query string.
+     *
+     * @param query The query string to search.
+     * @param limit The maximum number of entries to return.
+     * @return The list of matching BookmarkItems.
+     */
+    override suspend fun searchBookmarks(query: String, limit: Int): List<BookmarkItem> {
+        return withContext(scope.coroutineContext) {
+            reader.searchBookmarks(query, limit)
+        }
+    }
+
+    /**
+     * Adds a new bookmark tree node to a given node.
+     *
+     * @param newNode The node to add.
+     * @param position The optional position to add it in.
+     * @return The guid of the newly inserted bookmark item.
+     */
```

Comment:
Here and elsewhere where we end up using a `write` connection - can you please document Sync behaviour in the docstring? For an example, see how `PlacesHistoryStorage` documents sync behaviour. I think for bookmarks a simple note saying that `other clients will see this change if sync` should be enough.

Reply:
Done